### PR TITLE
2 session api

### PR DESCRIPTION
**Pull request by sravfeyn**
Adding back session-api with proper handling for internal APIs. I tried to find instances of internal API use by grepping for `api_dispatch_list`, `api_dispatch_detail` and `api_get_schema` (Tastypie urlnames). I found four endpoints that are used internally. I have added a `allow_internal` kwarg to fall back to session API for such endpoints (added tests as well). 

I am not 100% confident that this won't break something (I have looked around nginx logs but couldn't find any insights on API usage). If others aren't confident as well, we can do one of following
1. Do a phased roll out of this PR (staging, india, prod) similar to CSRF. 
2. Since we don't have insights into API usage, we can first do another PR that adds a function to log authorization methods for API requests. After we accumulate some logs, we can analyze them and fix if any issues and then release the session less API

@czue @millerdev 
buddy @kaapstorm

### CODE REVIEW

File: corehq/apps/api/tests.py [hunk 1]
```diff
@@ -1515,3 +1552,47 @@ def test_wrong_user_api_key(self):
 
         other_api_key.delete()
         other_user.delete()
+
+
+class InternalTestMixin(object):
```
Comment (sravfeyn) [inline corehq/apps/api/tests.py]: Tests for internal API endpoints

File: corehq/apps/api/resources/v0_1.py [hunk 2]
```diff
@@ -65,18 +71,32 @@ def _inner(req, domain, *args, **kwargs):
 
 class LoginAndDomainAuthentication(Authentication):
 
+    def __init__(self, allow_internal=False, *args, **kwargs):
+        """
+        allow_internal:
+            set this to True to allow session based access to this resource
```
Comment (millerdev) [inline corehq/apps/api/resources/v0_1.py]: The name `allow_internal` is confusing me. I wouldn't guess that it means "allow session-based access." What do you think of `allow_session_auth` instead?

**Comment by czue** [comment]
agree with @millerdev's feedback, but otherwise good to go from my side. your process for finding these makes sense to me, so if you've confirmed that the known usages are working with this change then I'm ok with that level of testing (though also happy to put under a staged release if you feel more comfortable with that)

**Comment by sravfeyn** [comment]
I have added tests for internal APIs have also tested location_internal API on local HQ, so these issues are fixed. Anyway, I will do a staged release. Currently keeping this on staging, will put it on India on Wednesday if no issues are found.

**Comment by millerdev** [comment]
@sravfeyn I added "do not merge" label. You can remove it when you're ready to have this merged. Uh... well I'm not sure how you're planning to do the staged rollout. It seems to me if we merge now it will get deployed everywhere today. Is there some other mechanism for phased rollout?

**Comment by millerdev** [comment]
👍 from me

**Comment by sravfeyn** [comment]
Sorry I never responded to the question ' how can we do phased rollout'. I realized we can't do this. So, I was okay having this just on staging for a while. It's been more than a week since it's on staging without any authorization problems for regular non API users. So I feel it should be good for this to go on prod now.

File: corehq/apps/api/tests.py [hunk 3]
```diff
@@ -1607,9 +1699,10 @@ def test_auth(self):
         new_user = WebUser.create(wrong_domain.name, 'test', 'testpass')
         new_user.save()
 
-        self.client.login(username='test', password='testpass')
-        response = self.client.get(self.single_endpoint(self.report_configuration._id))
-        self.assertEqual(response.status_code, 401)  # 401 is "Unauthorized"
+        response = self._assert_auth_get_resource(self.single_endpoint(self.report_configuration._id),
+                                                  username='test', password='testpass'
+        )
```
Comment (dimagimon) [inline corehq/apps/api/tests.py]: E124 closing bracket does not match visual indentation

**Comment by sravfeyn** [comment]
@millerdev

**Comment by millerdev** [comment]
kicked stuck build

**Pull request merged**